src/sage/graphs/matching.py

```diff
@@ -0,0 +1,1635 @@
+    # the graph is matching covered
+    if self.order() == 2:
+        return (True, None) if coNP_certificate else True
+
+    # A graph (without a self-loop) is matching covered if and only if the
+    # underlying simple graph is matching covered
+    G = self.to_simple()
+
+    if matching:
+        # The input matching must be a valid perfect matching of the graph
+        M = Graph(matching)
+        if any(d != 1 for d in M.degree()):
+            raise ValueError("the input is not a matching")
+        if not M.is_subgraph(G, induced=False):
+            raise ValueError("the input is not a matching of the graph")
+        if (G.order() != M.order()) or (G.order() != 2*M.size()):
+            raise ValueError("the input is not a perfect matching of the graph")
+    else:
+        # A maximum matching of the graph is computed
+        M = Graph(G.matching(algorithm=algorithm, solver=solver, verbose=verbose,
+                                integrality_tolerance=integrality_tolerance))
+
+        # It must be a perfect matching
+        if G.order() != M.order():
+            return (False, M.edges()[0]) if coNP_certificate else False
+
+    # Biparite graph:
+    #
+    # Given a connected bipartite graph G[A, B] with a perfect matching M.
+    # Construct a directed graph D from G such that V(D) := V(G) and
+    # for each edge in G direct the corresponding edge from A to B in D,
+    # if it is in M or otherwise direct it from B to A. The graph G is
+    # matching covered if and only if D is strongly connected.
+
+    if G.is_bipartite():
+        A, _ = G.bipartite_sets()
+        color = dict()
+
+        for u in G:
+            color[u] = 0 if u in A else 1
+
+        from sage.graphs.digraph import DiGraph
+        H = DiGraph()
+
+        for edge in G.edges():
+            u, v = edge[0], edge[1]
+
+            if color[u] == 1:
+                u, v = v, u
+
+            if M.has_edge(edge):
+                H.add_edge(u, v)
+            else:
+                H.add_edge(v, u)
+
+        # Check if H is strongly connected using Kosaraju's algorithm
+        def dfs(J, v, visited, p, parent, orientation):
+            stack = [(v, p)] # a stack of (vertex, parent)
+
+            while stack:
+                v, p = stack.pop()
+
+                if not visited[v]:
+                    visited[v] = True
+                    parent[v] = p
+
+                if orientation == 'in':
+                    for u in J.neighbors_out(v):
+                        if not visited[u]:
+                            stack.append((u, v))
+
+                elif orientation == 'out':
+                    for u in J.neighbors_in(v):
+                        if not visited[u]:
+                            stack.append((u, v))
+                else:
+                    raise ValueError('Unknown orientation')
+
+        root = next(H.vertex_iterator())
+
+        visited_in = {v: False for v in H}
```

Comment:
you could instead use a set, initially empty. Indeed, instead of tests like `if not visited[u]`, you can do `if u not in visited`.